pragma-node/src/utils/ws.rs

```diff
@@ -237,75 +332,152 @@ where
-                )
-                .await;
-                return Err(WebSocketError::MessageDecode(format!("{payload:?}")));
             }
+
             Message::Ping(_) => {
                 self.last_activity = std::time::Instant::now();
-                let _ = self.sender.send(Message::Pong(Default::default())).await;
+                let _ = self
+                    .send_sender
+                    .send(Message::Pong(Default::default()))
+                    .await;
             }
+
             Message::Pong(_) => {}
         }
         Ok(None)
     }
 
-    /// Send a message to the client.
-    pub async fn send_msg(&mut self, msg: String) -> Result<(), axum::Error> {
-        self.sender.send(Message::Text(msg.into())).await
+    /// Sends a message to the client after checking the rate limit.
+    pub async fn send_msg<T>(&mut self, message: T) -> Result<(), WebSocketError>
+    where
+        T: Sized + Serialize,
+    {
+        let message = serde_json::to_string(&message).map_err(WebSocketError::Serialization)?;
+
+        let message_size = message.len();
+        self.check_rate_limit(message_size).await?;
+        self.send_sender
+            .send(Message::Text(message.into()))
+            .await
+            .map_err(WebSocketError::SendError)?;
+
+        Ok(())
     }
 
-    /// Send an error message to the client without closing the channel.
-    pub async fn send_err(&mut self, err: &str) {
-        let err = json!({"error": err});
+    /// Sends an error message to the client.
+    pub async fn send_err(&self, msg: &str) {
+        let err = json!({"error": msg});
         let _ = self
-            .sender
+            .send_sender
             .send(Message::Text(err.to_string().into()))
             .await;
     }
 
-    /// Records a web socket metric.
+    /// Checks the rate limit for the given message size.
+    ///
+    /// If the limit is exceeded, it calls `handle_rate_limit_exceeded` to close the connection.
+    async fn check_rate_limit(&self, message_size: usize) -> Result<(), WebSocketError> {
+        let burst_size = NonZeroU32::new(message_size as u32)
+            .ok_or(WebSocketError::InternalError("Invalid message size".into()))?;
+
+        if self.rate_limiter.check_key_n(&self.ip_address, burst_size) != Ok(Ok(())) {
+            self.handle_rate_limit_exceeded().await?;
+            return Err(WebSocketError::RateLimitExceeded);
+        }
+        Ok(())
+    }
+
+    /// Handles the case when the rate limit is exceeded.
+    ///
+    /// Sends an error message to the client and closes the connection.
+    async fn handle_rate_limit_exceeded(&self) -> Result<(), WebSocketError> {
+        tracing::warn!(
+            subscriber_id = %self.id,
+            ip = %self.ip_address,
+            "Rate limit exceeded. Closing connection."
+        );
+
+        self.record_metric(metrics::Interaction::RateLimit, metrics::Status::Error);
+        self.send_err("Rate limit exceeded").await;
```

Comment:
wdyt about adding more info for the client e.g

```json
{
    "status": "error",
    "timestamp": "1664338190185908000",
    "error": "RATE_LIMIT_EXCEEDED",
    "data": {
        "retry_after": "4406000000"
    }
}
```

Reply:
done, just not sure about the `data` field so omitted for now, gonna check that in a follow-up